lib/Logger/LogContext.h

```diff
@@ -0,0 +1,559 @@
+  }
+
+  static inline constexpr char dummy[] = "dummy"; // only needed to calculate MinEntryCacheSize with MSVC
+  static constexpr std::size_t MinEntryCacheSize =
+      sizeof(EntryImpl<ValuesImpl<std::tuple<std::string>, dummy>>);
+
+ private:
+  static constexpr std::size_t MaxCachedEntries = 128;
+  CachedEntryAlloc* _cachedEntries = nullptr;
+  std::size_t _numCachedEntries = 0;
+};
+
+struct LogContext::ThreadControlBlock {
+  ThreadControlBlock() = default;
+  ThreadControlBlock(ThreadControlBlock const&) = delete;
+  ThreadControlBlock(ThreadControlBlock&&) = delete;
+  ThreadControlBlock& operator=(ThreadControlBlock const&) = delete;
+  ThreadControlBlock& operator=(ThreadControlBlock&&) = delete;
+
+  template <class Vals, const char... Keys[]>
+  LogContext::Entry* push(LogContext::ValueBuilder<Vals, Keys...>&& v);
+  LogContext::Entry* push(std::shared_ptr<LogContext::Values> v);
+
+  void pop(LogContext::Entry* entry) noexcept;
+
+ private:
+  friend class LogContext;
+  LogContext _logContext;
+  LogContext::EntryCache _entryCache;
+};
+
+struct LogContext::Accessor::ScopedValue {
+  explicit ScopedValue(std::shared_ptr<LogContext::Values> v) {
+    appendEntry(std::move(v));
+  }
+
+  template <class Vals, const char... Keys[]>
+  explicit ScopedValue(ValueBuilder<Vals, Keys...>&& v) {
+    using V = ValuesImpl<Vals, Keys...>;
+    appendEntry(V(std::move(v._vals)));
+  }
+
+  ScopedValue(ScopedValue const&) = delete;
+  ScopedValue& operator=(ScopedValue const&) = delete;
+  
+  ~ScopedValue();
+  
+  template <const char Key[], class Val>
+  static ScopedValue with(Val&& v) { return ScopedValue(LogContext::makeValue().with<Key>(std::forward<Val>(v))); }
+
+ private:
+  template <class V>
+  void appendEntry(V&& v) {
+    auto& local = LogContext::controlBlock();
+    auto e = local._entryCache.makeEntry<EntryImpl<V>>(std::forward<V>(v));
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+    _oldTail = e.get();
+#endif
+    local._logContext.pushEntry(std::move(e));
+  }
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+  Entry* _oldTail = nullptr;
+  std::thread::id _owningThread = std::this_thread::get_id();
+#endif
+};
+
+/// @brief sets the given LogContext as the thread's current context for the
+/// current scope; restores the previous LogContext upon destruction.
+struct LogContext::ScopedContext {
+  explicit ScopedContext(LogContext ctx) noexcept;
+  ~ScopedContext();
+
+ private:
+  std::optional<LogContext> _oldContext;
+};
+
+// These functions are deliberately defined inline in this header file to allow
+// more aggressive inlining for reduced overhead, even when compiling without IPO.
+
+// LogContext::EntryImpl
+
+template <class Vals>
+inline void LogContext::EntryImpl<Vals>::release(EntryCache& cache) noexcept {
+  if (sizeof(*this) <= LogContext::EntryCache::MinEntryCacheSize && !cache.isFull()) {
+    this->~EntryImpl();
+    cache.addToCache(this);
+  } else {
+    this->~EntryImpl();
+    ::operator delete(this);
+  }
```

Comment:
It looks like we are calling `this->~EntryImpl();` unconditionally:
```suggestion
  this->~EntryImpl();
  if (sizeof(*this) <= LogContext::EntryCache::MinEntryCacheSize && !cache.isFull()) {
    cache.addToCache(this);
  } else {
    ::operator delete(this);
  }
```